### PR TITLE
Move all visibility checking into AST-level function_checker, simplify that code a bit, and improve diagnostics.

### PR DESCRIPTION
**Pull request by brmataptos**
### Description

- move all visibility checking to AST and clean it up a  bit
- change `Known attribute ... position` warning to a neater `Attribute .. position` warning
- add FunctionData id_loc to allow pointing at function name in declaration for more concise error messages. abstract messages a bit in function_checker
- add 'inlined from' labels to diagnostics with labels

### Test Plan

Should just affect diagnostics.  Current tests suffice.

### CODE REVIEW

File: third_party/move/move-model/src/builder/module_builder.rs [hunk 1]
```diff
@@ -464,6 +463,7 @@ impl<'env, 'translator> ModuleBuilder<'env, 'translator> {
         );
         et.parent.parent.define_struct(
             et.to_loc(&def.loc),
+            // et.to_loc(&name.loc()),
```
Comment (vineethk) [inline third_party/move/move-model/src/builder/module_builder.rs:466]: remove leftover comment

File: third_party/move/move-compiler-v2/src/function_checker.rs [hunk 2]
```diff
@@ -53,11 +57,13 @@ pub fn check_for_function_typed_parameters(env: &mut GlobalEnv) {
...
 
 /// For all function in target modules:
-/// - check that non-inline function parameters do not have function type
-/// - check that calls to or from inline functions are accessible;
-///   - non-inline functions are handled in the VisibilityChecker pass, but
-///     inline functions will be either gone or inlined by then.
-pub fn check_access_and_use(env: &mut GlobalEnv) {
+///
+/// If `before_inlining`, then
+/// - check that all function calls involving inline funcitons are accessible;
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/function_checker.rs:62]:
```suggestion
/// - check that all function calls involving inline functions are accessible;
```

File: third_party/move/move-compiler-v2/src/function_checker.rs [hunk 3]
```diff
@@ -89,156 +93,175 @@ pub fn check_access_and_use(env: &mut GlobalEnv) {
...
+                        // Check visibility.
+
+                        // Same module is always visible
+                        let same_module = callee_env.module_env.get_id() == caller_module_id;
+                        let call_involves_inline_function =
+                            callee_env.is_inline() || caller_func.is_inline();
+
+                        // SKIP check if same_module or
+                        // if before inlining and the call doesn't involve inline function
+                        // OR if after inlining and call *does* involve an inline function
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/function_checker.rs:110]: After inlining, there should not be any inline function left, so this comment is a bit confusing. Is it possible that we are at "after inlining", and the call *does* involve an inline function?

File: third_party/move/move-compiler-v2/src/function_checker.rs [hunk 4]
```diff
@@ -89,156 +93,175 @@ pub fn check_access_and_use(env: &mut GlobalEnv) {
...
-                                                caller_module_name.display_full(env),
-                                                callee_env.module_env.get_full_name_str());
-                                            env.diag_with_labels(
-                                                Severity::Error,
-                                                &callee_env.get_loc(),
-                                                &msg,
-                                                call_details,
-                                            );
-                                        }
+                                        let why = format!(
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/function_checker.rs:134]: For the sake of consistency, pull this out into `not_friend_to_module_error()` or some such (mimicking `private_to_module_error()` below)?

File: third_party/move/move-compiler-v2/src/function_checker.rs [hunk 5]
```diff
@@ -89,156 +93,175 @@ pub fn check_access_and_use(env: &mut GlobalEnv) {
                     },
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/function_checker.rs:93]: Why do we need to record `private_funcs` if `!before_inlining`? (applies to the code above where we add to `private_funcs`, but I could not comment on those lines as they were not changed).

File: third_party/move/move-compiler-v2/src/function_checker.rs [hunk 6]
```diff
@@ -89,156 +93,175 @@ pub fn check_access_and_use(env: &mut GlobalEnv) {
...
-                            }
-                        );
-                        env.diag(Severity::Warning, &callee_loc, &msg);
+    if before_inlining {
+        // Check for Unused functions: private (or friendless public(friend)) funs with no callers.
+        let options = env
+            .get_extension::<Options>()
+            .expect("Options is available");
+        if options.warn_unused {
+            for callee in private_funcs {
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/function_checker.rs:176]: Are we missing out on warning a `public (friend)` function `foo` that is in a module with friends, but none of those friends call `foo`? Do we want in such cases as well?

File: third_party/move/move-compiler-v2/tests/testsuite.rs [hunk 7]
```diff
@@ -36,8 +36,8 @@ pub const EXP_EXT: &str = "exp";
 /// Configuration for a set of tests.
 #[derive(Default)]
 struct TestConfig {
-    /// Whether only type check should be run.
-    type_check_only: bool,
+    /// Whether stackless bytecode passes should be skipped.
+    skip_bytecode_passes: bool,
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/testsuite.rs:40]: I am not sure if this is accurate: this flag also skips file format generation, in addition to the stackless bytecode pipeline. Maybe `stop_before_sb_pipeline` or `stop_before_sb_gen` or similar?

File: third_party/move/move-compiler-v2/tests/visibility-checker/call_friend_only.exp [hunk 8]
```diff
@@ -1,13 +1,19 @@
 
 Diagnostics:
-error: `public(friend)` function `0xdeadbeef::M::id` cannot be called from `0xdeadbeef::Q::calls_bar` because `0xdeadbeef::Q` is not a friend of `0xdeadbeef::M`
-   ┌─ tests/visibility-checker/call_friend_only.move:46:9
+error: function `0xdeadbeef::M::bar` cannot be called from function `0xdeadbeef::Q::calls_bar` because module `0xdeadbeef::Q` is not a `friend` of `0xdeadbeef::M`
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/visibility-checker/call_friend_only.exp:3]: Maybe we should also move all these tests out from here and into `tests/checking` to keep up with the existing convention?

File: third_party/move/move-model/src/builder/model_builder.rs [hunk 9]
```diff
@@ -133,6 +133,7 @@ pub(crate) struct StructEntry {
 #[derive(Debug, Clone)]
 pub(crate) struct FunEntry {
     pub loc: Loc,
+    pub name_loc: Loc,
```
Comment (vineethk) [inline third_party/move/move-model/src/builder/model_builder.rs:136]:
```suggestion
    pub name_loc: Loc,  // location of just the function name
```

File: third_party/move/move-model/src/builder/model_builder.rs [hunk 10]
```diff
@@ -133,6 +133,7 @@ pub(crate) struct StructEntry {
 #[derive(Debug, Clone)]
 pub(crate) struct FunEntry {
     pub loc: Loc,
```
Comment (vineethk) [inline third_party/move/move-model/src/builder/model_builder.rs:135]:
```suggestion
    pub loc: Loc,  // location of the entire function span
```

**Review by vineethk — COMMENTED** [review]
I have not checked all the tests carefully whether the case of "visibility is okay before inlining but not okay after" has already been covered:

```
module 0xc0ffee::m {
    public inline fun foo() {
        priv();
    }

    fun priv() {}

    fun test() {
        foo(); // ok
    }
}

module 0xc0ffee::n {
    fun test() {
        0xc0ffee::m::foo(); // not ok
    }

}
```

If not, it would be nice to add this test. The error message produced on this one is really nice!

Looking good so far.

**Review by brmataptos — COMMENTED** [review]
Addressed Vineeth's comments already.  Forgot to send this response.

File: third_party/move/move-compiler-v2/src/function_checker.rs (hunk 6 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/src/function_checker.rs:166]: Good point.

**Review by vineethk — APPROVED** [review]

**Review by wrwg — APPROVED** [review]

**Pull request merged**